Ask java compiler to generate 1.6 compatible bytecode

isomarcte opened the pull request:
This will allow building this project with JDK 8 (the only version that is not EOL without an Oracle contract), while still supporting usage by those institutions that do have an aforementioned extended service contract.

* `pom.xml`
  * Added `target` to the `maven-compiler-plugin` and set to `1.6`. This should
    allow us to build with JDK 1.8 and still support 1.6 clients, keeping in
    mind we can not use Java Standard Library 1.8 exclusive symbols.
* `src/test/java/scala_maven/MiscTest.java`
  * Removed test that fails if you are using JDK > 7.

Note, I have successfully tested the plugin with against the invoker tests using JDK 1.6 using the output of this change.

davidB commented:
iirc, I keep the failing test because the split code was used somewhere (or to prevent its usage).
Need to check before merge.

Anyways, Thanks again. you can add your name in the contributor section if you like.

isomarcte commented:
A couple thoughts on this.

First some context, please skip if you are already aware of the semantics of the changes to regexs

## Context ##

The `split` change on JDK >= 1.8 is as follows. Normally, when a substring matches the regex at index 0 an empty `String` will be in the resulting array, on if there is a zero width match at index 0 then an empty `String` will _not_ be placed in the resulting.

```
scala> "fo".split("f")
res0: Array[String] = Array("", o)
```

However, if the match is zero width, then the empty `String` is _not_ placed in the Array. This is the change.

```
scala> "f".split("f") // Zero width because the split match is ""
res1: Array[String] = Array()
```

On JDK < 1.8 you'd end up with this.

```
scala> "f".split("f")
res1: Array[String] = Array("")
```

## Thoughts ##

1. We can't _really_ rely on the semantics of Standard Library code under any case. Having the tests case in `MiscTests` only checks this at compile time, but if someone chooses to use `scala-maven-plugin` running with Java 8 the check will do nothing since the standard library is runtime linked. i.e. if we currently require semantics for a particular JDK version we can either.
   1. Check `System.getProperties` and hard crash if we are run on the wrong JDK.
   2. Write forward/backward adapter.
2. If, in fact, we rely on the first `String` in an `Array` being empty under certain conditions, I'd argue that's a bad thing to rely upon.

However, I have been using this plugin with Java 8 for well over a year now without ever running into a bug related to this. Perhaps there was one in the past in one of the dependent libraries, but if it is still there it is in very uncommonly used code.

Thoughts?

davidB commented:
Agree about the issue about runtime jdk, the code is aged before 1.8 (when scala was incompatible with 1.8)
To be complete, I guess the change should replace String.split usage into the source, else Zinc support could be broken (may it is already broken).  I'll do the change (after merge)

davidB commented:
I remember now, I didn't enforce jdk6 at runtime, I created the test to not forgot to why I should not used String.split to split arguments in a cross jdk consistent way.
